### DISCUSSION TITLE
Looking for Advice on Customizing this Software for Mechanical Engineering Projects

### DISCUSSION THREAD

**Discussion by tabarsi1** [fougue/mayo#226] [Q&A]
Hey everyone,

First off, huge thanks for creating and sharing this fantastic open-source software

I'm a mechanical engineer looking to customize it to create a tool that aligns with my specific needs and ideas.

However, I'm not an expert in software development, so diving into this customization journey is proving to be a bit of a challenge. 

I'd really appreciate your insights on how to approach this. Any tips or steps to kick things off would be a game-changer for me!

Thanks a bunch for your time and guidance.

**Comment by HuguesDelorme** [comment]
Hi @tabarsi1
There's no that much documentation, however I recommend you have a look at this wiki page https://github.com/fougue/mayo/wiki/Design to get at least an understanding of the main source modules in Mayo
The most important classes in `base` module are `Application`, `Document`, `Property`, `IO::Reader`, `IO::Writer` and `IO::System`
For the graphics/GUI part have a look at `GraphicsScene` and `GuiDocument`
The app module is probably the most complex but basically `MainWindow`, `AppModule`, `Command `and `IAppContext ` are the root classes.
Any CAD file is loaded with an `IO::Reader` which imports geometry into a `Document`(data is organized with OpenCascade Data framework known as OCAF and XDE)
For each Document in Mayo there's a corresponding GuiDocument object that owns all the 3D graphics to be displayed
`MainWindow` is a page and command container. There are actually only two pages: _Home_ page(or the start page) and the _Documents_ page(where you can see the files you opened). The interface for `MainWindow` pages is `IWidgetMainPage`
Each application feature(new/open/import/export/...) is implemented in its own `Command` class

You can describe what you would like to customize in Mayo, and I'll be glad to can provide the relevants steps for the implementation

**Reply by itsivag** [reply]
@HuguesDelorme can you specify the steps for building the source it on windows as it is not that clear on the wiki especially the commands

**Reply by HuguesDelorme** [reply]
@itsivag 
Don't you want to install latest Mayo instead? Check the [Releases](https://github.com/fougue/mayo/releases) page

**Reply by itsivag** [reply]
Thanks for your reply but I'm looking forward to build it myself

**Reply by HuguesDelorme** [reply]
@itsivag
Please post here what commands you've tried and the errors you got

**Reply by itsivag** [reply]
while executing nmake all

```
Microsoft (R) Program Maintenance Utility Version 14.38.33130.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\bin\HostX86\x86\nmake.exe" -f Makefile.Release all

Microsoft (R) Program Maintenance Utility Version 14.38.33130.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        'C:\Qt\6.6.1\mingw_64\bin\uic.exe' ../src/app/dialog_about.ui -o ui_dialog_about.h
The filename, directory name, or volume label syntax is incorrect.
NMAKE : fatal error U1077: ''C:\Qt\6.6.1\mingw_64\bin\uic.exe' ../src/app/dialog_about.ui -o ui_dialog_about.h' : return code '0x1'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\bin\HostX86\x86\nmake.exe" -f Makefile.Release all' : return code '0x2'
Stop.  

```

the above error occurs i have executed the env.bat under occd

**Reply by HuguesDelorme** [reply]
@itsivag
Are you trying to build Mayo with mingw?
On Windows, Mayo is supposed to be built with Visual Studio, there's no guarantee mingw is supported(neved tried on my side)

**Reply by itsivag** [reply]
That wasn't intentional, i just had the path set to qt.exe and the only available directory is mingw is there any way to add msvc there or shall i have to install qt from start..

**Reply by HuguesDelorme** [reply]
@itsivag 
If you have Visual Studio installed then yes make sure you are using Qt built for Visual Studio
Qt installer is [here](https://www.qt.io/download-qt-installer)

**Reply by itsivag** [reply]
Tried qt installer but idk why they aren't interested in maintaining the open source server it always seem to fail

**Reply by HuguesDelorme** [reply]
@itsivag 
You ran the Qt installer and you got an error? What is the problem?

**Reply by itsivag** [reply]
The mirrors they provide don't seem to respond, every time I run it i encounter a download failed .... Network error i have tried switching my networks but nothing works and I searched it online and got to know many faced the issues

**Reply by HuguesDelorme** [reply]
@itsivag
The download errors happen after you validated the Qt components to install?

**Reply by itsivag** [reply]
Nope that doesn't even get across the initial 3 steps, like it asks for sign up and then on proceeding it returns a download failed error

**Reply by HuguesDelorme** [reply]
@itsivag
As a workaround you can try the prebuilt Qt binaries from this dedicated GitHub project:
https://github.com/martinrotter/qt-minimalistic-builds
Grab the "dynamic" variant in the [Releases](https://github.com/martinrotter/qt-minimalistic-builds/releases/tag/5.15.3) page

Never tried, but probably worth testing for you

**Reply by itsivag** [reply]
Thanks for it will try and catch you up🙏🏻

**Reply by itsivag** [reply]
```
C:\Users\itsivag\Qt Projects\mayo\build>qmake ..\mayo.pro
git version 2.39.1.windows.1
Project MESSAGE: Mayo version 0.9.0 debug
Project MESSAGE: Qt version 5.11.2
Project ERROR: Qt >= 5.14 is required but detected version is 5.11.2   
```

My Env :  C:\Qt2\5.15.2\msvc2019_64\bin
C:\OpenCASCADE-7.7.0-vc14-64\opencascade-7.7.0

**Reply by HuguesDelorme** [reply]
@itsivag
Your env is probably pointing on the Qt version provided by OpenCascade 3rd parties
Set again the Qt library with `set PATH=C:\Qt2\5.15.2\msvc2019_64\bin;%PATH%`

**Reply by itsivag** [reply]
That has been resolved but on running nmake all I encounter these errors. 


```
C:\Users\itsivag\Qt Projects\mayo\build>nmake all

Microsoft (R) Program Maintenance Utility Version 14.38.33130.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\bin\HostX64\x64\nmake.exe" -f Makefile.Release all

Microsoft (R) Program Maintenance Utility Version 14.38.33130.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        cl -c -nologo -Zc:wchar_t -FS -Zc:rvalueCast -Zc:inline -Zc:strictStrings -Zc:throwingNew -Zc:referenceBinding -Zc:__cplusplus /we4150 /std:c++17 -O2 -MD /Zi -std:c++17 -W3 -w34100 -w34189 -w44996 -w44456 -w44457 -w44458 -wd4577 -wd4467 -EHsc -DUNICODE -D_UNICODE -DWIN32 -D_ENABLE_EXTENDED_ALIGNED_STORAGE -DWIN64 -DQT_DISABLE_DEPRECATED_BEFORE=0x050F00 -DQT_IMPLICIT_QFILEINFO_CONSTRUCTION -DNOMINMAX -DOCCT_HANDLE_NOCAST -D_OCC64 -DNDEBUG -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_WINEXTRAS_LIB -DQT_GUI_LIB -DQT_CORE_LIB -I..\..\mayo -I. -Isrc\app -I..\src\app -I..\src\3rdparty -IC:\Qt2\5.15.2\msvc2019_64\include -IC:\Qt2\5.15.2\msvc2019_64\include\QtWidgets -IC:\Qt2\5.15.2\msvc2019_64\include\QtWinExtras -IC:\Qt2\5.15.2\msvc2019_64\include\QtGui -IC:\Qt2\5.15.2\msvc2019_64\include\QtANGLE -IC:\Qt2\5.15.2\msvc2019_64\include\QtCore -Irelease -I. -I/include -IC:\Qt2\5.15.2\msvc2019_64\mkspecs\win32-msvc -Forelease\ @C:\Users\itsivag\AppData\Local\Temp\nm1FC2.tmp
application.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\application_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Application.hxx': No such file or directory
application_item.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\application_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Application.hxx': No such file or directory
application_item_selection_model.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\application_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Application.hxx': No such file or directory
bnd_utils.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\bnd_utils.h(10): fatal error C1083: Cannot open include file: 'Bnd_Box.hxx': No such file or directory
brep_utils.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\occ_brep_mesh_parameters.h(9): fatal error C1083: Cannot open include file: 'Standard_Version.hxx': No such file or directory
caf_utils.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\caf_utils.h(9): fatal error C1083: Cannot open include file: 'Standard_Version.hxx': No such file or directory
document.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\application_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Application.hxx': No such file or directory
document_tree_node.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\document_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Document.hxx': No such file or directory
geom_utils.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\geom_utils.h(10): fatal error C1083: Cannot open include file: 'gp_Pnt.hxx': No such file or directory
io_reader.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\document_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Document.hxx': No such file or directory
io_system.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\application_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Application.hxx': No such file or directory
label_data.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\caf_utils.h(9): fatal error C1083: Cannot open include file: 'Standard_Version.hxx': No such file or directory
math_utils.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\math_utils.h(9): fatal error C1083: Cannot open include file: 'gp_Dir.hxx': No such file or directory
mesh_access.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\mesh_access.h(13): fatal error C1083: Cannot open include file: 'Quantity_Color.hxx': No such file or directory
mesh_utils.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\occ_handle.h(9): fatal error C1083: Cannot open include file: 'Standard_Handle.hxx': No such file or directory
occ_progress_indicator.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\tkernel_utils.h(9): fatal error C1083: Cannot open include file: 'Quantity_Color.hxx': No such file or directory
occ_static_variables_rollback.cpp
..\src\base\occ_static_variables_rollback.cpp(9): fatal error C1083: Cannot open include file: 'Interface_Static.hxx': No such file or directory
occt_enums.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\occt_enums.h(10): fatal error C1083: Cannot open include file: 'Aspect_HatchStyle.hxx': No such file or directory
point_cloud_data.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\point_cloud_data.h(9): fatal error C1083: Cannot open include file: 'Graphic3d_ArrayOfPoints.hxx': No such file or directory
property_builtins.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\math_utils.h(9): fatal error C1083: Cannot open include file: 'gp_Dir.hxx': No such file or directory
Generating Code...
Compiling...
property_value_conversion.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\filepath_conv.h(12): fatal error C1083: Cannot open include file: 'TCollection_AsciiString.hxx': No such file or directory
quantity.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\math_utils.h(9): fatal error C1083: Cannot open include file: 'gp_Dir.hxx': No such file or directory
string_conv.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\string_conv.h(9): fatal error C1083: Cannot open include file: 'NCollection_UtfString.hxx': No such file or directory
task_manager.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\math_utils.h(9): fatal error C1083: Cannot open include file: 'gp_Dir.hxx': No such file or directory
text_id.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\application_ptr.h(9): fatal error C1083: Cannot open include file: 'TDocStd_Application.hxx': No such file or directory
tkernel_utils.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\tkernel_utils.h(9): fatal error C1083: Cannot open include file: 'Quantity_Color.hxx': No such file or directory
triangulation_annex_data.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\triangulation_annex_data.h(11): fatal error C1083: Cannot open include file: 'Quantity_Color.hxx': No such file or directory
unit_system.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\math_utils.h(9): fatal error C1083: Cannot open include file: 'gp_Dir.hxx': No such file or directory
xcaf.cpp
C:\Users\itsivag\Qt Projects\mayo\src\base\math_utils.h(9): fatal error C1083: Cannot open include file: 'gp_Dir.hxx': No such file or directory
Generating Code...
NMAKE : fatal error U1077: 'cl -c -nologo -Zc:wchar_t -FS -Zc:rvalueCast -Zc:inline -Zc:strictStrings -Zc:throwingNew -Zc:referenceBinding -Zc:__cplusplus /we4150 /std:c++17 -O2 -MD /Zi -std:c++17 -W3 -w34100 -w34189 -w44996 -w44456 -w44457 -w44458 -wd4577 -wd4467 -EHsc -DUNICODE -D_UNICODE -DWIN32 -D_ENABLE_EXTENDED_ALIGNED_STORAGE -DWIN64 -DQT_DISABLE_DEPRECATED_BEFORE=0x050F00 -DQT_IMPLICIT_QFILEINFO_CONSTRUCTION -DNOMINMAX -DOCCT_HANDLE_NOCAST -D_OCC64 -DNDEBUG -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_WINEXTRAS_LIB -DQT_GUI_LIB -DQT_CORE_LIB -I..\..\mayo -I. -Isrc\app -I..\src\app -I..\src\3rdparty -IC:\Qt2\5.15.2\msvc2019_64\include -IC:\Qt2\5.15.2\msvc2019_64\include\QtWidgets -IC:\Qt2\5.15.2\msvc2019_64\include\QtWinExtras -IC:\Qt2\5.15.2\msvc2019_64\include\QtGui -IC:\Qt2\5.15.2\msvc2019_64\include\QtANGLE -IC:\Qt2\5.15.2\msvc2019_64\include\QtCore -Irelease -I. -I/include -IC:\Qt2\5.15.2\msvc2019_64\mkspecs\win32-msvc -Forelease\ @C:\Users\itsivag\AppData\Local\Temp\nm1FC2.tmp' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\bin\HostX64\x64\nmake.exe" -f Makefile.Release all' : return code '0x2'
Stop.
```

**Reply by HuguesDelorme** [reply]
@itsivag 
Path to OpenCascade seems not correct
Make sure to run this command before qmake(and set PATH for Qt bin):
`call C:\OpenCASCADE-7.7.0-vc14-64\opencascade-7.7.0\env.bat`

**Reply by itsivag** [reply]
Yup I have run it and nothing seemed to change will try once again

**Reply by itsivag** [reply]
After few attempts i was able to set the qt path and run it but it returns this build error

```
Generating Code...
        link /NOLOGO /DYNAMICBASE /NXCOMPAT /OPT:REF /INCREMENTAL:NO /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF /SUBSYSTEM:WINDOWS "/MANIFESTDEPENDENCY:type='win32' name='Microsoft.Windows.Common-Controls' version='6.0.0.0' publicKeyToken='6595b64144ccf1df' language='*' processorArchitecture='*'" /VERSION:0.9 /MANIFEST:embed /OUT:release\mayo.exe @C:\Users\itsivag\AppData\Local\Temp\nmEAC7.tmp
release\document_tree_node_properties_provider.obj : fatal error LNK1112: module machine type 'x86' conflicts with target machine type 'x64'
NMAKE : fatal error U1077: 'link /NOLOGO /DYNAMICBASE /NXCOMPAT /OPT:REF /INCREMENTAL:NO /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF /SUBSYSTEM:WINDOWS "/MANIFESTDEPENDENCY:type='win32' name='Microsoft.Windows.Common-Controls' version='6.0.0.0' publicKeyToken='6595b64144ccf1df' language='*' processorArchitecture='*'" /VERSION:0.9 /MANIFEST:embed /OUT:release\mayo.exe @C:\Users\itsivag\AppData\Local\Temp\nmEAC7.tmp' : return code '0x458'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\bin\HostX64\x64\nmake.exe" -f Makefile.Release all' : return code '0x2'
Stop.
```

**Reply by HuguesDelorme** [reply]
@itsivag 
There's a 32/64b conflict, I guess there's a mixin between x86 and x64 libs during linkage.
Make sure you installed Qt x64 binaries

**Reply by itsivag** [reply]
Yup worked on it and had my build successfully but the following error occurs while trying to run the debug and release versions:
TKBRep.dll 
TKBin.dll 
TKBinXCAF.dll
TKBinL.dll

all these were not found!!

**Reply by HuguesDelorme** [reply]
Try to run `mayo.exe` from the command prompt where you setup your build environment(`env.bat` and so on)

**Reply by itsivag** [reply]
On running it from cmd it shows 

`The procedure entry point?hitButton@QPushButton@@MEBA_NAEBVQPoint@@@Z could not be located in the dynamic link library C:\users\itsivag\Qt Projects\mayo\build\release\mayo.exe`

**Reply by HuguesDelorme** [reply]
@itsivag 
Make sure from that command prompt you typed:
````c++
call C:\OpenCASCADE-7.7.0-vc14-64\opencascade-7.7.0\env.bat
set PATH=C:\Qt2\5.15.2\msvc2019_64\bin;%PATH%
````
Then run `mayo.exe`

**Reply by itsivag** [reply]
Thank you. I'm able to run it and anyway to run it without cmd and is there any community channel in discord so i could connect with you

**Reply by HuguesDelorme** [reply]
Glad to see it finally works for you!
There is unfortunately no community channel for Mayo

**Reply by itsivag** [reply]
Any way to direct message you

**Reply by itsivag** [reply]
Hey there, I'm trying to compile the project in qt creator. Any idea on how to include the headers of open cascade in qt creator.

**Reply by HuguesDelorme** [reply]
@itsivag
The simplest but least flexible solution is to run QtCreator from a command prompt where you previously typed `call C:\OpenCASCADE-7.7.0-vc14-64\opencascade-7.7.0\env.bat`
`mayo.pro` will find needed OpenCascade paths from the `CSF_*` environement variables set in `env.bat`

**Reply by itsivag** [reply]
After doing that the libs are linked successfully but building it directly from qt creator makes it encounter the same errors as opened from file explorer any permanent solution on that.

The errors were:
TKBRep.dll
TKBin.dll
TKBinXCAF.dll
TKBinL.dll

**Reply by HuguesDelorme** [reply]
@itsivag 
Make sure you checked that option:
![image](https://github.com/fougue/mayo/assets/939132/d05a1d1b-4a35-43a3-8a91-96207c75e29b)

**Reply by itsivag** [reply]
![Screenshot (69)](https://github.com/fougue/mayo/assets/85468732/848fb0a7-e738-470a-af9f-73731d9af6f8)

It's checked by default .. still facing the issues

**Reply by HuguesDelorme** [reply]
@itsivag 
Check also the environment is on "System Environment":
![image](https://github.com/fougue/mayo/assets/939132/efb6168b-f825-4cec-8b3d-e614052ffa89)
And Mayo should run fine

**Reply by itsivag** [reply]
Hey there, this is a Mainwindow and i need to create a sphere and attach it to a occ viewer and display it in the main window

```
#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <AIS_InteractiveContext.hxx>
#include <AIS_Shape.hxx>
#include <BRepPrimAPI_MakeSphere.hxx>
#include <V3d_View.hxx>
#include <Aspect_Window.hxx>
#include <OpenGl_GraphicDriver.hxx>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    

        // Create the 3D viewer
    Handle_OpenGl_GraphicDriver glGraphicDriver = new OpenGl_GraphicDriver(new Aspect_DisplayConnection);
    Handle_V3d_Viewer viewer = new V3d_Viewer(glGraphicDriver);
    viewer->SetDefaultLights();
    viewer->SetLightOn();
    
    Handle(AIS_InteractiveContext) context = new AIS_InteractiveContext(viewer);
    
    // Create a sphere and display it
    BRepPrimAPI_MakeSphere sphereMaker(10);
    TopoDS_Solid aShape = sphereMaker.Solid();
    Handle(AIS_Shape) aShapePrs = new AIS_Shape(aShape);
    context->Display(aShapePrs, AIS_Shaded, 0, true);
      
     
}

MainWindow::~MainWindow()
{
    delete ui;
}
```

I dont' have any idea how to proceed beyond this.. could you please help me make this and btw is there any good articles/tuts/docs for occt and qt integration?

**Comment by HuguesDelorme** [comment]
@itsivag 
See this discussion: https://github.com/fougue/mayo/discussions/81

**Reply by itsivag** [reply]
but my context was outside of mayo it's a separate project

**Reply by itsivag** [reply]
Hey, what does ApplicationItemSelectionModel actually do? in the comments i tend to see that it  

> Keeps track of the items selected in an Application object

but what does that mean?

**Reply by itsivag** [reply]
and btw i wanna add a custom functionality to it where when pressing right mouse button i have a generate a sphere on the surface of the mesh where the cursor is clicked and i know right mouse click mapping is to pan the model and i need to overwrite it.  I have looked #81 and tried to reproduce the functionality as you stated, I wanted to add the same to generate a shape on the cursor coordinate (where the mouse right click is done).

**Reply by HuguesDelorme** [reply]
@itsivag
Have a look at signal `V3dViewController::signalMouseButtonClicked` in `src/graphics/v3d_view_controller.h `
Add a slot function(eg a lambda) to that signal and check that the button clicked is `Aspect_VKeyMouse_RightButton`
The view controller is owned by `WidgetGuiDocument`, you can also extend the [existing slot](https://github.com/fougue/mayo/blob/4e76085e22c653cb61f66c2dff68da099a32739a/src/app/widget_gui_document.cpp#L198)

**Reply by itsivag** [reply]
on the mainWindow where can i find the code to create these new doc,open doc and recent files i found them under widget_home_files.h but i can't find the UI part for it.

**Reply by HuguesDelorme** [reply]
@itsivag
Each command in Mayo(ie which appears in menu bar) is implemented as `Mayo::Command` sub-class
Look at these references:
https://github.com/fougue/mayo/blob/4e76085e22c653cb61f66c2dff68da099a32739a/src/app/commands_api.h#L60
https://github.com/fougue/mayo/blob/4e76085e22c653cb61f66c2dff68da099a32739a/src/app/commands_file.h#L22
See how to add new commands in Mayo:
https://github.com/fougue/mayo/discussions/81

**Reply by itsivag** [reply]
so, i just wanna modify the look of the dashboard.
![Screenshot (76)](https://github.com/fougue/mayo/assets/85468732/f13fc2ab-8a02-46bc-8a48-32e16bcee837)

So, i just need the path to the file where these UI is created, I could find where the UI widgets are connected with the logic (widget_main_home.cpp).

**Reply by HuguesDelorme** [reply]
@itsivag
The code you are looking for is here: 
https://github.com/fougue/mayo/blob/8e3302fe93d7092913f2e061bc32866b26be3501/src/app/widget_home_files.cpp#L37

The class `HomeFilesModel` provides a Qt "grid" model of the items displayed in the home page(or dashboard as you mentionned)

**Reply by itsivag** [reply]
Thank you

**Reply by itsivag** [reply]
I wanted to change the bg of listView_OpenedDocuments that shows the model tree of the document to have a white bg i have tried using a stylesheet both in qt designer and on the code on widget_main_control.cpp but i couldn't change it is there any other way?

**Reply by HuguesDelorme** [reply]
@itsivag
Try to run mayo.exe with `--theme classic` arguments
If it you like the theme it can be forced at startup in main.cpp:
https://github.com/fougue/mayo/blob/8e3302fe93d7092913f2e061bc32866b26be3501/src/app/main.cpp#L546
-> `globalTheme.reset(createTheme("classic"));`

**Reply by itsivag** [reply]
I want to move the tetrahedron and the controls on the ooct view to bottom right where to edit their position and other attributes

![Screenshot (88)](https://github.com/fougue/mayo/assets/85468732/df3fc42b-fbbc-48be-8a5a-f4c17cb7078f)

**Reply by HuguesDelorme** [reply]
@itsivag
The view cube itself can be moved in the bottom right with `GuiDocument::setViewTrihedronCorner(Aspect_TOTP_RIGHT_LOWER)`
This can be called in `src/app/widget_gui_document.cpp` circa L209
There's an issue with placement of the widget controls at that corner, this needs to be fixed
If you can propose a patch(pull request) that would be great!
Note: the relevant function is `WidgetGuiDocument::layoutViewControls()`

**Reply by itsivag** [reply]
sure, I'll try to resolve and create a PR

**Reply by itsivag** [reply]
btw can't be the controls placed at the top left regardless of the trihedron's position.
Anyway, but with trial and error the following values have been subtracted from the coords so that it doesn't overflow the screen.

```
     switch (m_guiDoc->viewTrihedronCorner()) {
            case Aspect_TOTP_LEFT_UPPER:
                return { ctrlXOffset, viewCubeBndSize + margin };
            case Aspect_TOTP_RIGHT_UPPER:
                return { this->width() - viewCubeBndSize - ctrlXOffset - 75, viewCubeBndSize + margin };
            case Aspect_TOTP_LEFT_LOWER:
                return { ctrlXOffset, this->height() - viewCubeBndSize - margin - ctrlHeight };
            case Aspect_TOTP_RIGHT_LOWER:
                return { this->width() - viewCubeBndSize + ctrlXOffset - 75,
                         this->height() - viewCubeBndSize - margin - ctrlHeight };
            default:
                return { margin, margin };
            }
```

**Reply by HuguesDelorme** [reply]
Yes the widget controls can be placed anywhere you want, just make it not depend on viewTrihedronCorner in `layoutViewControls()` function

**Reply by itsivag** [reply]
![Screenshot (98)](https://github.com/fougue/mayo/assets/85468732/18dfba9f-8a70-4bfa-a953-10b47f722c2e)
![Screenshot (97)](https://github.com/fougue/mayo/assets/85468732/72f9ec47-73fb-423e-859c-22905acb07ef)
![Screenshot (96)](https://github.com/fougue/mayo/assets/85468732/adb80b41-beaf-44e6-9c90-b966f5de2635)
![Screenshot (95)](https://github.com/fougue/mayo/assets/85468732/0cd782d0-af74-4011-a172-8d9ac1d5cc33)

can I know is this the functionality you expected?

**Reply by HuguesDelorme** [reply]
@itsivag
Not really, the view controls are supposed to always be placed "under" the 3D view cube
But that's not what you want, so you can forget about the PR and layout them as you wish for your project

**Reply by itsivag** [reply]
You can specify what's your requirement I'll try to contribute.
